executor/src/witgen/jit/symbolic_expression.rs

```diff
@@ -283,13 +297,71 @@ impl<T: FieldElement, V: Clone> SymbolicExpression<T, V> {
             self.clone()
         } else {
             SymbolicExpression::BinaryOperation(
-                Rc::new(self.clone()),
+                Arc::new(self.clone()),
                 BinaryOperator::IntegerDiv,
-                Rc::new(rhs.clone()),
+                Arc::new(rhs.clone()),
                 None,
             )
         }
     }
+
+    /// Transform the variables inside the expression using the given function
+    pub fn map_variables<V2, F: FnMut(&V) -> V2>(&self, f: &mut F) -> SymbolicExpression<T, V2> {
+        match self {
+            SymbolicExpression::Concrete(n) => SymbolicExpression::Concrete(*n),
+            SymbolicExpression::Symbol(s, rc) => SymbolicExpression::Symbol(f(s), rc.clone()),
+            SymbolicExpression::BinaryOperation(lhs, op, rhs, rc) => {
+                SymbolicExpression::BinaryOperation(
+                    Arc::new(lhs.map_variables(f)),
+                    op.clone(),
+                    Arc::new(rhs.map_variables(f)),
+                    rc.clone(),
+                )
+            }
+            SymbolicExpression::UnaryOperation(op, expr, rc) => SymbolicExpression::UnaryOperation(
+                op.clone(),
+                Arc::new(expr.map_variables(f)),
+                rc.clone(),
+            ),
+            SymbolicExpression::BitOperation(expr, op, n, rc) => SymbolicExpression::BitOperation(
+                Arc::new(expr.map_variables(f)),
+                op.clone(),
+                *n,
+                rc.clone(),
+            ),
+        }
+    }
+
+    /// Convert to an RPNExpression
+    pub fn to_rpn(&self) -> RPNExpression<T, V> {
+        let mut elems = Vec::new();
+        self.to_rpn_inner(&mut elems);
+        RPNExpression { elems }
+    }
+
+    fn to_rpn_inner(&self, elems: &mut Vec<RPNExpressionElem<T, V>>) {
```

Comment:
Are you worried about machine stack size or is the RPN transformation a performance optimization?

Reply:
from some profiling, the "evaluate expression" code was a hot spot... RPN was quite a bit faster.
I'm not sure if its just due to avoiding recursion or because the whole expression is kept in a vec (avoiding pointers).

Reply:
Yeah, thought, about the same, locality might be a thing.

Could you move this function to the interpreter, though? I think SymbolicExpression should not care that there is a thing called RPNExpressionElem.